File: core/src/main/java/org/elkd/core/consensus/RaftLeaderState.java
package org.elkd.core.consensus;

import com.google.common.base.Preconditions;
import com.google.common.util.concurrent.ListenableFuture;
import io.grpc.stub.StreamObserver;
import org.apache.log4j.Logger;
import org.elkd.core.cluster.ClusterMessenger;
import org.elkd.core.consensus.messages.*;
import org.elkd.core.statemachine.SetStateMachineCommand;

import javax.annotation.Nonnull;
import java.util.Timer;
import java.util.TimerTask;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

class RaftLeaderState implements RaftState {
  private static final Logger LOG = Logger.getLogger(RaftLeaderState.class.getName());
  private static final int TIMEOUT_MS = 1000;
  private static final ExecutorService EXECUTOR = Executors.newSingleThreadExecutor();

  private Timer mMonitor;

  private final Raft mRaft;
  private final ClusterMessenger mClusterMessenger;

  /* package */ RaftLeaderState(@Nonnull final Raft raft,
                                @Nonnull final ClusterMessenger clusterMessenger) {
    mRaft = Preconditions.checkNotNull(raft, "raft");
    mClusterMessenger = clusterMessenger;
  }

  @Override
  public void on() {
    LOG.info("ready");

    /* DEMO CODE */
    mRaft.getClusterConnectionPool().iterator().forEachRemaining(node -> {
      final AppendEntriesRequest request = AppendEntriesRequest.builder(0, 0, 0, mRaft.getClusterSet().getSelfNode().getId(), 0)
          .withEntry(Entry.builder("jonny").withCommand(new SetStateMachineCommand("x", "5")).build())
          .build();
      final ListenableFuture<AppendEntriesResponse> future = mClusterMessenger.appendEntries(node, request);
      future.addListener(() -> {
        try {
          LOG.info(future.get());
        } catch (Exception e) {
          e.printStackTrace();
        }
      }, EXECUTOR);

      final RequestVoteRequest request2 = RequestVoteRequest.builder(0, mRaft.getClusterSet().getSelfNode().getId(), 0, 0).build();
      final ListenableFuture<RequestVoteResponse> future2 = mClusterMessenger.requestVote(node, request2);
      future2.addListener(() -> {
        try {
          LOG.info(future2.get());
        } catch (Exception e) {
          e.printStackTrace();
        }
      }, EXECUTOR);
    });

    restartMonitor();
  }

  @Override
  public void off() {
    LOG.info("offline");
    stopMonitor();
  }

  @Override
  public void delegateAppendEntries(final AppendEntriesRequest appendEntriesRequest,
                                    final StreamObserver<AppendEntriesResponse> responseObserver) {
    responseObserver.onNext(AppendEntriesResponse.builder(102, true).build());
    responseObserver.onCompleted();
    restartMonitor();
  }

  @Override
  public void delegateRequestVote(final RequestVoteRequest requestVoteRequest,
                                  final StreamObserver<RequestVoteResponse> responseObserver) {
    responseObserver.onNext(RequestVoteResponse.builder(0, true).build());
    responseObserver.onCompleted();
    restartMonitor();
  }

  private void restartMonitor() {
    stopMonitor();
    mMonitor = new Timer(false);
    mMonitor.schedule(new TimerTask() {
      @Override
      public void run() {
        mRaft.transition(RaftFollowerState.class);
      }
    }, TIMEOUT_MS, TIMEOUT_MS);
  }

  private void stopMonitor() {
    if (mMonitor != null) {
      mMonitor.cancel();
    }
  }
}

File: core/src/main/java/org/elkd/core/consensus/RaftLeaderState.kt
package org.elkd.core.consensus

import io.grpc.stub.StreamObserver
import org.elkd.core.consensus.messages.AppendEntriesRequest
import org.elkd.core.consensus.messages.AppendEntriesResponse
import org.elkd.core.consensus.messages.RequestVoteRequest
import org.elkd.core.consensus.messages.RequestVoteResponse
import org.elkd.core.consensus.replication.Replicator
import org.elkd.core.log.LogChangeReason
import org.elkd.core.log.commands.AppendCommand
import org.elkd.core.runtime.client.command.Command

class RaftLeaderState(private val raft: Raft) : RaftState {
  private var replicator: Replicator? = null

  override fun on() {
    /* for test sake, append a new entry to the logger here so we have something to replicate */
    broadcastLeaderChange()
    replicator = Replicator(raft, LeaderContext(raft.clusterSet.nodes, raft.log.lastIndex))
    replicator?.start()
  }

  override fun off() {
    /* Force-stop the replication process - we must honor the transition */
    replicator?.stop()
  }

  override val supportedOps = setOf(OpCategory.PRODUCE, OpCategory.COMMAND, OpCategory.CONSUME)

  override fun delegateAppendEntries(request: AppendEntriesRequest,
                                     stream: StreamObserver<AppendEntriesResponse>) {
    stream.onNext(AppendEntriesResponse.builder(raft.raftContext.currentTerm, false).build())
    stream.onCompleted()
  }

  override fun delegateRequestVote(request: RequestVoteRequest,
                                   stream: StreamObserver<RequestVoteResponse>) {
    /* If term > currentTerm, Raft will always transition to Follower state. model received
       here will only be term <= currentTerm so we can defer all logic to the consensus delegate.
     */
    stream.onNext(RequestVoteResponse.builder(raft.raftContext.currentTerm, false).build())
    stream.onCompleted()
  }

  private fun broadcastLeaderChange() {
    raft.logCommandExecutor.execute(AppendCommand.build(
        Command.builder(Command.Type.LEADER_CHANGE) {
          arg("node", raft.clusterSet.localNode.id)
        }.asEntry(raft.raftContext.currentTerm),
        LogChangeReason.REPLICATION
    ))
  }
}
